src/wallet/wallet.cpp

```diff
@@ -1118,21 +1118,10 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFlushOnClose)
     bool fUpdated = false;
     if (!fInsertedNew)
     {
-        // Merge
-        if (!wtxIn.hashUnset() && wtxIn.hashBlock != wtx.hashBlock)
-        {
-            wtx.hashBlock = wtxIn.hashBlock;
-            fUpdated = true;
-        }
-        // If no longer abandoned, update
-        if (wtxIn.hashBlock.IsNull() && wtx.isAbandoned())
-        {
-            wtx.hashBlock = wtxIn.hashBlock;
-            fUpdated = true;
-        }
-        if (wtxIn.nIndex != -1 && (wtxIn.nIndex != wtx.nIndex))
-        {
-            wtx.nIndex = wtxIn.nIndex;
+        if (wtxIn.tx_state.state != wtx.tx_state.state) {
+            wtx.tx_state.state = wtxIn.tx_state.state;
+            wtx.tx_state.nIndex = wtxIn.tx_state.nIndex;
+            wtx.tx_state.hashBlock = wtxIn.tx_state.hashBlock;
             fUpdated = true;
         }
```

Comment:
Done

Reply:
(re: https://github.com/bitcoin/bitcoin/pull/16624#discussion_r315293956)

I don't like these new asserts. I think they could be triggered if:

- a wallet tx is confirmed in a block
- the wallet is shut down
- the block chain re-orgs and the tx is included in a different block
- the wallet file is loaded on a sync'ed node

All this code is trying to do is update an existing wallet tx's `Confirmation` status with the new `Confirmation` status. I think we can just change it to:

`if (wtxIn.tx_state.state != wtx.tx_state.state || wtxIn.tx_state.hashBlock != wtx.tx_state.hashBlock || wtxIn.tx_state.nIndex != wtx.tx_state.nIndex) {`

Reply:
If these new asserts get triggered it would mean we have bugs in higher level code, but yes aborting maybe too strong given that this kind of bugs could be due to some likely API misuse in RPCs ? What's about logging them with a warning ?

I'm not in favor of proposed alternative as it may cause hidden inconsistencies if this aforementioned high level bugs are present

Reply:
I don't feel strongly, but I think asserts are way to go because we should not allow bugs and uncertainty in wallet event processing code. It's hard for me to imagine a scenario where it would be a win to use defensive programming in a place like this.

Reply:
@ryanofsky - do you agree that the scenario I describe above would trigger the assert?

Reply:
Oh, I didn't realize https://github.com/bitcoin/bitcoin/pull/16624#discussion_r315883326 was describing a specific scenario, not just a list of possible events. I could see that scenerio triggering these asserts, and making the change you suggested an improvement and bugfix. So it is likely some change is needed here, and the suggested one seems ok.

More ideally, though, it'd be really nice to have a functional test that triggers these asserts, and instead of hiding the problem of an unconfirmed transaction being left in the CONFIRMED state after a reorg, there would be code on startup that marks reorged transactions UNCONFIRMED before rescanning (could basically be the same code from #15931 that looks up block heights on startup).

Nice catch though (assuming this a bug)!

Reply:
> I didn't realize #16624 (comment) was describing a specific scenario

sorry - changed my unordered list to an ordered list and updated wording to make this clearer.

Reply:
> there would be code on startup that marks reorged transactions UNCONFIRMED before rescanning

Yes - I think if the wallet rescans from block at height x, it should first change all transactions that are CONFIRMED at height >x to UNCOFIRMED.

Reply:
I let asserts and move some bits from #15931 to transition status to UNCONFIRMED if blockHash not anymore in chain. Also added a wallet_reorgsrestore test with aforementioned scenario, without new code it triggers asserts